common/autoware_auto_perception_rviz_plugin/src/object_detection/predicted_objects_display.cpp

```diff
@@ -25,161 +28,238 @@ namespace object_detection
 {
 PredictedObjectsDisplay::PredictedObjectsDisplay() : ObjectPolygonDisplayBase("tracks")
 {
-  std::thread worker(&PredictedObjectsDisplay::workerThread, this);
-  worker.detach();
+  num_threads_property = new rviz_common::properties::IntProperty(
+    "Num Threads", 4, "Num of Threads to use", this);  // by default, use 4 threads
+  max_num_threads = 8;            // hard code the number of threads to be created
+  num_threads = max_num_threads;  // this will be soon modified in the workerThread
+
+  for (int ii = 0; ii < max_num_threads; ++ii) {
+    threads.emplace_back(std::thread(&PredictedObjectsDisplay::workerThread, this));
+  }
+
+  // Initialize the semaphores, which signify the completion of the marker creation tasks.
+  task_completed_semaphores = reinterpret_cast<sem_t *>(malloc(max_num_threads * sizeof(sem_t)));
```

Comment:
> task_completed_semaphores is never released - memory leak

task_completed_semaphores are expected to be released in the destructor of the plugin.

```c
  ~PredictedObjectsDisplay()
  {
    {
      std::unique_lock<std::mutex> lock(queue_mutex);
      should_terminate = true;
    }
    mutex_condition.notify_all();
    for (std::thread & active_thread : threads) {
      active_thread.join();
    }
    threads.clear();

    for (int ii = 0; ii < max_num_threads; ++ii) {
      sem_close(&task_completed_semaphores[ii]);
    }
    ObjectPolygonDisplayBase<autoware_auto_perception_msgs::msg::PredictedObjects>::
      ~ObjectPolygonDisplayBase<autoware_auto_perception_msgs::msg::PredictedObjects>();
  }
```

> use std::mutex.

So is it better to implement a semaphore class using std::mutex rather than using the semaphore library?
https://stackoverflow.com/questions/4792449/c0x-has-no-semaphores-how-to-synchronize-threads

This seems to be so fundamental and I would expect this to be implemented in a common/ library in Autoware but not in a plugin.

Reply:
@ralwing Could you check this?